### ISSUE TITLE
OmitDeep - when omitting multiple optional fields final type goes wrong

### ISSUE THREAD

**Issue by vilola** [sindresorhus/type-fest#966]
### Bug description

When OmitDeep multiple optional fields, final type is not correct. See example.

### Repro

https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBDAnmApnA3geRMGARFFMAXzgDMoIQ4ByJVAWjJQGcYaBuAWAChf60mMDGAQAdgEMANgDlWMFABM4AXgy84muBIBc6nlsNwARgH496DUesC9bKMDEBzbgetxir93ADG5-d6GtnD2ji5W1p4RRor+lm6BCMgodjAOzl7uUQkergCQTv4ARKHORQA0vHkAFsWlThW82fzJcHJsSqoBWrrd1sYW0d7B9ZmRY0Y+gznDyanp4TNa2d6K04maI2lhE4Yry-lOeiXbZZU8Ncf1jTzNPAD093AACpTGUiggEiI+cGASsBaqDgBCIQhE4mkABV5F1sLhQWAADzg0SSWTyJTlOBFCQAOmMRTgAB8cfifEUAHwTR5wAB6pl4vFpAHVcNU4GIIPAIMI0dJyMAUFJFCw4LgxQB3aAAazCQLQiJhbDhOHwhGR7QUimxuIJRNJeop1KZDyeUN5cA+ADdhXB4fBpVA5c44JL2XBeRD0Qr7XzIVJlfA1A6Uf70VqsTiGiScdUqdFaQzeEA

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/sindresorhus) so you can upvote and help fund this issue.
- The funding will be given to active contributors.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/sindresorhus/type-fest/issues/966">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/sindresorhus/type-fest/issues/966/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/sindresorhus/type-fest/issues/966/pledge.svg">
</picture>
</a>

**Comment by Emiyaaaaa** [comment]
You mean the result has extra undefined?

**Comment by vilola** [comment]
> You mean the result has extra undefined?
> 
> Try set `exactOptionalPropertyTypes: true` in tsconfig.

exactOptionalPropertyTypes did not do anything. 

Example of an issue once more:

```
type Info1 = {
    clientUserId?: string;
    surname: string;
    userInfo: {
        pep: string;
        number: string;
        uselessInfo: {
            foo: string;
        };
        uselessField?: string;
    };
};

type UsefulInfo1 = OmitDeep<Info1, "clientUserId" | "userInfo.uselessField">;
// type UsefulInfo1 = {
//     surname: string;
//     userInfo: {
//         pep: string;
//         number: string;
//         uselessInfo: {
//             foo: string;
//         };
//         uselessField?: string;
//     };
// } | {
//     clientUserId?: string;
//     surname: string;
//     userInfo: {
//         number: string;
//         pep: string;
//         uselessInfo: {
//             foo: string;
//         };
//     };
// }
```


When all of the omitted fields are in the original type possibly undefined, the final type is not correct - it is a tuple of types each with one of the key omitted, non where are omitted all of given. 
Interestingly when at least one of the given keys is required, it works fine and the optionals are omitted as well and the final type is correct - one object with fields without the ones mentioned in the OmitDeep

**Comment by vilola** [comment]
This is also not expected behaviour 

```
type Info1 = {
    clientUserId?: string;
    surname: string;
    userInfo: {
        pep: string;
        number: string;
        uselessInfo: {
            foo: string;
        };
        uselessField?: string;
    };
};

type Info2 = SetRequired<Info1, "clientUserId">;

type UsefulInfo1 = OmitDeep<Info2, "clientUserId" | "userInfo.uselessField">;
// type UsefulInfo1 = {
//     surname: string;
//     userInfo:
//         | {
//               pep: string;
//               number: string;
//               uselessInfo: {
//                   foo: string;
//               };
//               uselessField?: string;
//           }
//         | {
//               number: string;
//               pep: string;
//               uselessInfo: {
//                   foo: string;
//               };
//           };
// };

```

**Comment by vilola** [comment]
Still not working 

```

type Info1 = {
    clientUserId: string;
    surname: string;
    userInfo: {
        pep: string;
        number: string;
        uselessInfo: {
            foo: string;
        };
        uselessField?: string;
    };
};

type UsefulInfo1 = OmitDeep<Info1, "clientUserId" | "userInfo.uselessField">;
// type UsefulInfo1 = {
//     surname: string;
//     userInfo: {
//         pep: string;
//         number: string;
//         uselessInfo: {
//             foo: string;
//         };
//         uselessField?: string;
//     } | {
//         number: string;
//         pep: string;
//         uselessInfo: {
//             foo: string;
//         };
//     };
// }
```

**Comment by vilola** [comment]
Working - all required

```
type Info1 = {
    clientUserId: string;
    surname: string;
    userInfo: {
        pep: string;
        number: string;
        uselessInfo: {
            foo: string;
        };
        uselessField: string;
    };
};

type UsefulInfo1 = OmitDeep<Info1, "clientUserId" | "userInfo.uselessField">;
// type UsefulInfo1 = {
//     surname: string;
//     userInfo: {
//         pep: string;
//         number: string;
//         uselessInfo: {
//             foo: string;
//         };
//     };
// }
```

**Comment by vilola** [comment]
Working - nested required

```
type Info1 = {
    clientUserId?: string;
    surname: string;
    userInfo: {
        pep: string;
        number: string;
        uselessInfo: {
            foo: string;
        };
        uselessField: string;
    };
};

type UsefulInfo1 = OmitDeep<Info1, "clientUserId" | "userInfo.uselessField">;
// type UsefulInfo1 = {
//     surname: string;
//     userInfo: {
//         pep: string;
//         number: string;
//         uselessInfo: {
//             foo: string;
//         };
//     };
// }
```

**Comment by som-sm** [comment]
@vilola Sorry, but I'm unable to understand what's _not_ working exactly? 

This seems correct to me:
<img width="1059" alt="image" src="https://github.com/user-attachments/assets/bb34081a-18f1-4854-b390-ef817cacca99" />

[Playground](https://tsplay.dev/N571BW)

**Comment by vilola** [comment]
You are correct, it is fixed in v4.30.0 from last week, I see now after upgrade 👍
